Sources/NostrSDK/CustomEmoji.swift

```diff
@@ -0,0 +1,73 @@
+//
+//  CustomEmoji.swift
+//
+//
+//  Created by Terry Yiu on 11/11/23.
+//
+
+import Foundation
+
+/// [NIP-30 Custom Emoji](https://github.com/nostr-protocol/nips/blob/master/30.md)
+public class CustomEmoji: CustomEmojiValidating, Equatable {
+    public static func == (lhs: CustomEmoji, rhs: CustomEmoji) -> Bool {
+        lhs.isEqual(to: rhs)
+    }
+
+    /// A name given for the emoji, which MUST be comprised of only alphanumeric characters and underscores.
+    let shortcode: String
+
+    /// A a URL to the corresponding image file of the emoji.
+    let imageUrl: URL
+
+    init?(shortcode: String, imageUrl: URL) {
+        self.shortcode = shortcode
+        self.imageUrl = imageUrl
+
+        if !isValidShortcode(shortcode: shortcode) {
```

Comment:
Use `guard` and put this at the top of the initializer.

Reply:
I'm getting this error for some reason. Is there any way I can solve it without needing to rename my variables?

```
Constant 'self.shortcode' used before being initialized
```

Reply:
Seems like I could just keep the guard at the bottom after the constants get initialized, though not ideal.

Reply:
Ah. It's because the function `isValidShortcode` requires that `self` exists, so it wants the required properties to be initialized when the function is called. Kinda weird, but I think it's worth it to keep the function separated from the type in that way.